Scene Editor GUI

Matt95 opened the pull request:
An initial version of the GUI based scene editor.

Currently includes:
- Positioning and resizing via drag and drop, mouse wheel and toolbar.
- Templated positioning
- Switching between random and manual position
- Lock/unlock aspect ratio
- Handle different media types.
- Reset to best fit while preserving aspect ratio.
- Autosave

There were some partially completed items that I have not included to get to a working version (hopefully) free of major bugs, these can probably be added back in at another time.
- selectable scene aspect ratio 
- multiple item positioning

TODO's:
- fix grid resizing issue (fixed aspect content and freeform grid means user can resize incorrectly).

Notes:
- Some old content with partially specified position can't be loaded without making assumptions so the editor displays a warning instead.
- I have a scene - "sizetest" that acts as a test with different media types, positioning etc. 

Please test it out and let me know if there are any issues.

aaronpcz commented on src/js/components/pages/adjustable-grid.jsx:
```diff
@@ -58,6 +59,7 @@ var RespGrid = React.createClass({
     getComponent: function (item) {
         var self = this;
 
+            //TODO: should we not be using layout constants here as well? MK.
```
Yep Layout Constants should be used here.

Stormgun commented on src/js/components/pages/scene-editor-gui.jsx:
```diff
@@ -0,0 +1,793 @@
+        this.setState({width: this.SceneLayoutArea.offsetWidth, height: this.SceneLayoutArea.offsetHeight})
+    },
+
+    componentWillReceiveProps: function (nextProps) {
```
This lifecycle function can be used to get the resizing of the component in gear based on the updated size of the parent.

Stormgun commented on src/js/components/pages/scene-editor-gui.jsx:
```diff
@@ -0,0 +1,793 @@
+        GridStore.addChangeListener(this._onFocusChange)
+
+        //this is the first point where rendered size of area can be captured
+        this.setState({width: this.SceneLayoutArea.offsetWidth, height: this.SceneLayoutArea.offsetHeight})
```
At first, I thought that we should use the this.refs , but that is for dom reference, while you are referencing a class of rectangle.

Stormgun commented on src/js/components/pages/scene-editor-gui.jsx:
```diff
@@ -0,0 +1,793 @@
+
+    _onFocusChange: function () {
+
+        focusedMediaObject = GridStore.getFocusedMediaObject();
```
Undeclared variables in a few places, they need to be declared so that they do not bubble up outside of the function, if this is not done carefully it can cause variable overrides and unexpected bugs.

Matt95 replied:
Will fix this

Stormgun commented on src/js/components/pages/scene-editor-gui.jsx:
```diff
@@ -0,0 +1,793 @@
+            this.getImageSize(self.state.mediaObject.url, function(imgWidth, imgHeight) {
+    
+                scaleX = imgWidth/self.state.width;
+                scaleY = imgHeight/self.state.height;
```
Values are unused and the math seems switched 2 lines down - needs to be removed and declared

Matt95 replied:
Yep top 2 lines can be removed

Stormgun commented on src/js/components/pages/scene-editor-gui.jsx:
```diff
@@ -0,0 +1,793 @@
+                default:
+                    //unsupported media type
+                    mediaTypeSupported = false
+                    mediaRepresentation = null
```
Unused variable mediaRepresentation -> this is the only place in the code that this is used, unsure on its purpose

Matt95 commented:
Should now be cleared up, removed a few bugs in the process, "select a scene" message now working correctly and fixed a missing state variable that was causing errors. Also eliminated cyclic saving due to scene update.  Templates for quadrents and fullscreen now remove rotation

aaronpcz commented:
Awesome we are going to process all PRs tonight and create a new uber version of this project ready for tomorrow... will let you know how it goes.
